RYCRX varform

Cryoris opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add it in the CHANGELOG file under Unreleased section.
⚠️ If your pull request fixes an open issue, please link to the issue.

✅ I have added the tests to cover my changes.
✅ I have updated the documentation accordingly.
✅ I have read the CONTRIBUTING document.
-->

### Summary
Add a new variational form called RYCRX, which, heuristically, has a high expressibility. 


### Details and comments
In arxiv.org/abs/1905.10876 the authors propose this newly implemented variational form with a relatively low number of parameters for the high expressibility of the underlying Hilbert space. 
We generalised the proposed circuit and called this default behaviour `'circular'` entanglement.

It is also possible to use `'full'` or `'linear'` entanglement, provide an own `entangler_map` and to skip unentangled qubits.

A test based on the VQE test has been implemented and was passed, the docstring documentation and changelog have been updated.

chunfuchen commented:
From my understanding, this PR including two things
1. A new way to define how to entangle qubits and each layer could have different entanglement
2. Using parameterized CRX (or CRZ) as entanglement rather than using CX and CZ.

For point 1, I think it is more flexible if we update the `get_entangler_map` method to support other entangler maps, and that method could generate one entangler map for each layer instead of one entangler map for all layers. In this way, we can re-use such entangler map for all variational form we had.

On the other hand, for point 2, I think this variational form is a variant of RY variational form with CRX as the two-qubit gate. (You can call current RY as RYCX or RYCZ.)
I will prefer to make this CRX as an option when constructing the variational form and then it will be easy to extend for other variational forms as we can also have RYRZCRX by just changing how to entangle two qubits.

Cryoris commented:
I see the advantages of your suggestions and point 2 is also straightforward to implement.

Concerning point 1, what do you think would be the optimal way to support individual `entangler_map`s for the layers?

One possibility would be to add the optional keyword `layer_id` (or similar). This would have the advantage that no existing code using linear or full entanglement would need amendments.

We could of course also add a `num_layers` parameter and if its larger than 1 return a list of entangler maps. However the first one seems more elegant and versatile if one would want the entangler map of a specific layer (no need to generate all before that).
```
def get_entangler_map(map_type, num_qubits, layer_id=0):
    ret = []
    if num_qubits > 1:
        if map_type == 'full':
            ret = [[i, j] for i in range(num_qubits) for j in range(i + 1, num_qubits)]
        elif map_type == 'linear':
            ret = [[i, i + 1] for i in range(num_qubits - 1)]
        elif map_type = 'circular':
            # use layer_id to determine proper entanglement
        else:
            raise ValueError("map_type only supports 'full' or 'linear' type.")
    return ret

chunfuchen commented:
I was more thinking on the way that we provide the number of blocks (or layers) to that method but I think your solution (`layer_id`) is better. Btw, could we call `layer_id` as `offset`? from my understanding, it is the offset where the circle started.

Cryoris commented:
Yes, `offset` is a clearer name. Alright, then I'll go about implementing the changes. 
Thanks for the constructive suggestions!

dongreenberg commented:
@Cryoris This is a great addition. Where is this ansatz mentioned in that paper? I didn't spot it.

Cryoris commented:
@dongreenberg See circuit 14 on page 8 of [this paper](https://arxiv.org/pdf/1905.10876.pdf).
Then we extracted the underlying structure and generalised it for more layers (note that in our interpretation circuit 14 is shown with 2 layers) and more qubits. 

I've attached a screenshot from an explanatory notebook I've prepared on this.
![ExplanationCircuit14](https://user-images.githubusercontent.com/5978796/59435287-8b93de00-8ded-11e9-89f8-812d6d3768e0.png)

dongreenberg commented:
Ah, thanks. I think you linked a different paper in your opening comment above.

Cryoris commented:
Yes, I did, thanks for the hint! I updated the opening comment.

Cryoris commented:
@chunfuchen I've integrated the RYCRX varform as we discussed above. I've added the option `skip_final_ry` to skip the final RY layer and reproduce the desired circuit from the paper. 
If new entangler maps with individual entanglements per block are implemented, all one has to change in `ry.py` is one line from:
```
if self._entanglement == 'circular':
```
to
```
if self._entanglement in ['circular', 'my_new_entanglement']:
```
This if is currently used to not call `get_entangler_map` multiple times if we only have `'full'` or `'linear'` entanglement. It could also be removed for more the sake of elegance.. 😉 

Please let me know if the code is fine now!

dongreenberg commented:
@chunfuchen sorry to chime in late... but I think breaking out the rycrx varform as its own class is more easily discovered and understood by practitioners than an entangler map option inside of an existing varform. I could see why we wouldn't want an explosion of classes for every possible variant, but we're adding this specifically because it was shown to be high-performing. 

Alternate, completely opposing opinion: the possible explosion reflects that the varform classes can be made much more flexible. The relationship between rycrx and ry is similar to the relationship between ry and ryrz. Maybe we just need one varform object for shingled/hw-efficient ansatze, and let the user configure the 1q and 2q gates - e.g.: `my_vf = ShingledVarForm(1q='ryrx', 2q='crz', entangler_map='linear')`. That would cover ry, ryrz, swaprz, rycrx, and I think all of those in the paper above. I can break out the discussion into a separate bug.

chunfuchen commented on qiskit/aqua/components/variational_forms/ry.py:
```diff
@@ -138,17 +155,33 @@ def construct_circuit(self, parameters, q=None):
+
             for src, targ in self._entangler_map:
-                if self._entanglement_gate == 'cz':
+                if self._entanglement_gate.lower() == 'cz':
```
I think `lower()` is not needed since the schema checking will raise an error if you do not provide the lower case of `cz`, `cx` and `crx`.

Cryoris replied:
Ah yes, I'll change it, thanks!

chunfuchen commented:
@Cryoris Sorry, I should comment this at the first place, I checked out the paper and the paper mentioned that when they made the circuit deeper (increase L), they simply repeat the blocks but I do not find how they change the entanglement for the repeated blocks? Do you mind pointing me out where the paper mentioned that?

On the other hand, regarding the name of entanglement (`circular`), for my understanding, this entanglement is beyond the normal circular since the even and odd blocks have an opposite way to define the pair of [control_qubit, target_qubit]. 
Since this entanglement is designed based on the paper circuit14, could we name it based on the paper and also add the reference to the docstring? The name could be something like `sim-circ14` or others. 
Thanks.

chunfuchen commented:
@dongreenberg I am not sure whether or not we should expose it as another variational form class. I think if this variation is better than the original RY variational form, we can change the default of RY variational form with CRX. 

Regarding having a shingled variational form, could we open another issue on that? If the hardware-efficient variational form is always designed in a similar way, like current RYRZ, RY, and this RYCRX, it is good to have a common underlying implementation to reduce our effort when we would like to experiment with different types of variational forms.

Zoufalc commented:
@chunfuchen I wouldn't necessarily change the default of RY to using CRX b.c. it leads to a non-negligible increase in the number of parameters of the VarForm. Thus, the optimizers need to update more parameters. There is certainly a trade-off between expressibility of the VarForm and complexity of the optimization process for e.g. VQE. 
I agree with @dongreenberg that we should either put the new VarForm into a separate class or generalize to a shingled VarForm. As mentioned above the new VarForm does not only use a CRX gate for entanglement but also a different entangler map layout and comes without the last RY layer.

Cryoris commented:
@chunfuchen As you correctly pointed out, there is a difference in the circuit 14 of the paper and our implementation. This is on purpose. 
* In the paper circuit 14 consists of 2 RY layers and 2 CRX entanglement layers. With increasing depth this full block gets repeated and no entanglement gate shifted. 
* In our implementation one block is made up of 1 RY layer and 1 CRX entanglement layer. Per new layer the entanglements get shifted. Per block we have half the number of parameters than in the original circuit 14.
* Circuit 14 can be represented with our implementation if the number of layers is set to 2.

As mentioned in the PR description our VarForm is a generalisation of circuit 14. It has two main improvements:
1. The length and therefore number of parameters is more precisely adjustable which is important for the possibilities we have with current devices, and 
2. the entanglement is varies more and doesn't simply get repeated every block, thus probably providing a better representation of the underlying Hilbert space. 

Furthermore it wasn't clear from the paper how the underlying circuit structure is to be extended to more than 4 qubits. Although measurements for 6 and 8 qubits are provided (in their appendix) I couldn't find any comment on how this was implemented. How we now implemented the extension to an arbitrary number of qubits is in my opinion as general as possible while still reproducing the "original" circuit 14 when using L=2 layers and 4 qubits.

So eventually it's a VarForm **based on** circuit 14 of this paper, but not the circuit itself.
It's a generalised version addressing the issues of the original circuit.

This maybe wasn't very obvious from the PR, please excuse me, I should've stated this more clearly.

chunfuchen commented:
@Cryoris Thanks for the explanation. It is great that this implementation generalizes on circuit 14 of the paper, which is good! We can apply for more qubits and different depths.

chunfuchen commented:
@Zoufalc Yes, it is not necessary to change the default. It is difficult to define which variational is **better** since it involves so many perspectives. 

I am not sure to create a new variational form because, in this case, should we specialize all variational forms when we have more? and the name of all variational forms should be changed. like current RY should be changed to RYCX for the consistency.
With the current modification, you can get the previous RYCRX by configuring the new RY correctly.
If this new variational form providing more evidence that it is useful for most cases, I think we should specialize it as a separate class. 

I agree with that providing unified underlying implementation and thus, users can freely configure the gates on 1q gate and 2q gates.

Cryoris commented:
I also believe that @dongreenberg's suggestion of shingled variational forms is a suitable design for the generic nature of variational forms. This should definitely be implemented in the long term. 

However for the not-so-experienced Aqua user who might be new to variational forms in general this will be quite confusing. To maintain usability throughout all kinds of users Aqua has I think it will be inevitable to provide wrappers of the kind that we already have: `RY` or `RYRZ`. But since they can simply call the shingled varform internally this implementation pattern is still quite clean.

Since the design for variational forms is a general Aqua topic and since this discussion might go on for a bit, it might be best discussed separately (as @dongreenberg proposed above already).
Also it would be great to allow access to the RYCRX varform soon.

Therefore, if you agree with this, we can maybe merge this PR once we implemented a more suitable name, and then work on this new problem without having this one still open.

Regarding the name: After some consideration referring to 'circuit 14' might be a bit ominous, since the average user probably won't look up the paper and there's not much information in this name. Some acronym or abbreviation of it's features (shifted circular with alternating entanglement) might be better. We're thinking of something.

chunfuchen commented:
@Cryoris Thanks. I think before we get more evidence that such RYCRX achieves better performance, e.g. fewer iterations or better quantum state than existed variational forms, we should just leave it inside current RY. 

I think that adding a class wrapper might just confuse users if users do not know which one is better for their problems; on the other hand, if users had the knowledge on using CRX is better, then they can easily configure it. 

All changes look good. But one more thing about the test file name. Do you mind renaming the `test_rycrx.py` to `test_ry.py` since it is under RY variational form now.

Cryoris commented:
Sure 👍

chunfuchen commented:
@Cryoris Do you mind helping update the documentation at https://github.com/Qiskit/qiskit?

Cryoris commented:
@chunfuchen Sure, the entry is already prepared, I'll create a PR. Thanks!
